### PR TITLE
http_proxy: support multi-addresses listening

### LINKED ISSUE #629: Support multiple ports binding
Our use case is that we run a proxy that listens on multiple ports using the same upstream.
Those ports are mentioned in [this doc](https://docs.saucelabs.com/secure-connections/sauce-connect/advanced/specifications/#supported-browsers-and-ports).

Listeners on those additional ports are optional - if the proxy fails to bind to one of the ports, it just logs an error and continues (it happens with ephemeral ports).

Would it be possible to add this functionality to Forwarder?

### PR DESCRIPTION
**Pull request by Choraden**
This patch adds multi-addresses listener. 

Fixes #629

### CODE REVIEW

File: bind/flag.go [hunk 1]
```diff
@@ -93,6 +93,12 @@ func HTTPProxyConfig(fs *pflag.FlagSet, cfg *forwarder.HTTPProxyConfig, lcfg *lo
 	HTTPServerConfig(fs, &cfg.HTTPServerConfig, "", forwarder.HTTPScheme, forwarder.HTTPSScheme)
 	LogConfig(fs, lcfg)
 
+	fs.StringSliceVar(&cfg.ExtraAddresses, "extra-address", cfg.ExtraAddresses,
+		"<host:port,...>"+
+			"Additional addresses to listen on. "+
+			"The flag can be specified multiple times to add multiple addresses. "+
+			"Bind failure is not critical, the proxy will continue to run. ")
```
Comment (waggledans) [inline bind/flag.go:100]: can we make it configurable? In my request, bind failures shouldn't be critical (and it should be a default) but it feels like it should be more generic

Comment (Choraden) [reply bind/flag.go:100]: We can always make it configurable. The question is do we need it right now? If not, I'd opt for not adding it, to keep the flags simple.

Comment (waggledans) [reply bind/flag.go:100]: fair point. Configurability is not needed right now and this functionality is not battle-tested yet, let's see how it goes.

**Comment by mmatczuk** [comment]
I'd suggest to extend the port syntax we use in the current flag

```
flag := [<host>]:<port_expr>
port_expr := <port>|<port-range>,...
port-range := <port>-<port>
```

Example:

--address localhost:50,60,70-80

Maybe the port range is too much and can be ignored.

The logic should be that we shall always be able to listen to one or more ports.

File: net.go [hunk 2]
```diff
@@ -89,24 +91,64 @@ type ListenerCallbacks interface {
 
 	// OnTLSHandshakeError is called after a TLS handshake errors out.
 	OnTLSHandshakeError(*tls.Conn, error)
+
+	// OnBindError is called when a listener fails to bind to an address.
+	OnBindError(address string, err error)
```
Comment (mmatczuk) [inline net.go:96]: Should go before TLS things.

File: net.go [hunk 3]
```diff
@@ -89,24 +91,64 @@ type ListenerCallbacks interface {
...
 }
 
+// Listener is a multi-address listener with TLS support, rate limiting and callbacks.
+// The first address in the list must bind successfully and is used as the listener address.
 type Listener struct {
-	Address          string
+	Addresses        []string
```
Comment (mmatczuk) [inline net.go:102]: Listener changes should be in a separate commit.

File: net.go [hunk 4]
```diff
@@ -89,24 +91,64 @@ type ListenerCallbacks interface {
...
 	Log              log.Logger
 	TLSConfig        *tls.Config
 	HandshakeTimeout time.Duration
 	ReadLimit        int64
 	WriteLimit       int64
 	Callbacks        ListenerCallbacks
 
-	listener net.Listener
+	ls        []net.Listener
```
Comment (mmatczuk) [inline net.go:110]: Why not listeners

File: net.go [hunk 5]
```diff
@@ -89,24 +91,64 @@ type ListenerCallbacks interface {
...
+	l.acceptCh = make(chan acceptResult, len(l.Addresses))
+	l.closeCh = make(chan struct{})
+
+	for i, addr := range l.Addresses {
+		ll, err := l.listen(addr)
+		if err != nil {
+			if l.Callbacks != nil {
+				l.Callbacks.OnBindError(addr, err)
+			}
+			if i == 0 {
```
Comment (mmatczuk) [inline net.go:133]: I think that when using more than one port they are all equal.

File: net.go [hunk 6]
```diff
@@ -116,13 +158,37 @@ func (l *Listener) Listen() error {
 		ll = ratelimit.NewListener(ll, wl, rl)
 	}
 
-	l.listener = ll
-	return nil
+	return ll, nil
+}
+
+func (l *Listener) acceptLoop(ll net.Listener) {
```
Comment (mmatczuk) [inline net.go:164]: I'd use `errgroup` with anonymous function instead of creating this and closeCh.

File: net_test.go [hunk 7]
```diff
@@ -70,3 +122,6 @@ func (m *mockListenerCallback) OnTLSHandshakeError(_ *tls.Conn, err error) {
 	}
 	m.donec <- struct{}{}
 }
+
```
Comment (mmatczuk) [inline net_test.go:125]: Order.

**Comment by mmatczuk** [comment]
As for the integration, the main part should be in the listener and it should go to the HttpServer.

**Comment by mmatczuk** [comment]
The `--address` and `--optional-addresses` is a usability mess it immediately raises questions like why would you ever need that.

All of that would not be a problem if we focused on file based configuration only - say we would have listener config object like Envoy. Then you could add optional field to it. We also have protocol and bunch of other stuff that make adding multiple addresses imbalance all of that.

Maybe we are solving wrong problem and this should be done on a script level?

**Comment by waggledans** [comment]
> Maybe we are solving wrong problem and this should be done on a script level?

What do you have in mind? The problem is described in https://github.com/saucelabs/forwarder/issues/629

> The --address and --optional-addresses is a usability mess

I'd rather have some way to assign properties to listeners, not sure how to make it work with CLI. 

>  it immediately raises questions like why would you ever need that.

hmm... https://github.com/saucelabs/forwarder/issues/629 ?

**Comment by waggledans** [comment]
> What that even means?, Why we should so easily give up on the ports, What's the purpose?

All these questions indicate that we need to improve Sauce Labs docs if it's not clear from the description.  Let's have this discussion elsewhere (we should have had it BEFORE the implementation).

> Why not use a script to launch a bunch of forwarders instead 

Absolutely not. It's 20-30 ports, I don't think so.

**Comment by mmatczuk** [comment]
Let's split the work and move the port forwarding to separate feature https://github.com/saucelabs/forwarder/issues/669
